apstools/plans/xpcs_mesh.py

```diff
@@ -0,0 +1,237 @@
+                        return False
+                # if there are any extra arguments, check that they have a default
+                else:
+                    if p.kind is p.VAR_KEYWORD or p.kind is p.VAR_POSITIONAL:
+                        continue
+                    if p.default is p.empty:
+                        return False
+
+            return True
+
+        def _verify_nd_step(sig):
+            if len(sig.parameters) < 3:
+                return False
+            for name, (p_name, p) in zip_longest(["detectors", "step", "pos_cache"], sig.parameters.items()):
+                # this is one of the first 3 positional arguements, check that the name matches
+                if name is not None:
+                    if name != p_name:
+                        return False
+                # if there are any extra arguments, check that they have a default
+                else:
+                    if p.kind is p.VAR_KEYWORD or p.kind is p.VAR_POSITIONAL:
+                        continue
+                    if p.default is p.empty:
+                        return False
+
+            return True
+
+        if sig == inspect.signature(bps.one_nd_step):
+            pass
+        elif _verify_nd_step(sig):
+            # check other signature for back-compatibility
+            pass
+        elif _verify_1d_step(sig):
+            # Accept this signature for back-compat reasons (because
+            # inner_product_scan was renamed scan).
+            dims = len(list(cycler.keys))
+            if dims != 1:
+                raise TypeError(
+                    "Signature of per_step assumes 1D trajectory " "but {} motors are specified.".format(dims)
+                )
+            (motor,) = cycler.keys
+            user_per_step = per_step
+
+            def adapter(detectors, step, pos_cache):
+                # one_nd_step 'step' parameter is a dict; one_id_step 'step'
+                # parameter is a value
+                (step,) = step.values()
+                return (yield from user_per_step(detectors, motor, step))
+
+            per_step = adapter
+        else:
+            raise TypeError(
+                "per_step must be a callable with the signature \n "
+                "<Signature (detectors, step, pos_cache)> or "
+                "<Signature (detectors, motor, step)>. \n"
+                "per_step signature received: {}".format(sig)
+            )
+    pos_cache = defaultdict(lambda: None)  # where last position is stashed
+    cycler = utils.merge_cycler(cycler)
+    motors = list(cycler.keys)
+
+    @bpp.stage_decorator(list(detectors) + motors)
+    @bpp.run_decorator(md=_md)
+    def scan_until_completion():
+        """
+        Scanning until the total number of required collection points is achieved
+        """
+        if predeclare:
+            yield from bps.declare_stream(*motors, *detectors, name="primary")
+        iterations = 0
+        while iterations < number_of_collection_points:
```

Comment:
I believe the outer `while` loop may be eliminated using `itertools.cycle` on top of the `cycler()`.  See this SO post: https://stackoverflow.com/a/23416519/1046449

Like this:

```py
        iterations = 0
        for step in itertools.cycle(cycler):
            yield from per_step(detectors, step, pos_cache)
            iterations += 1
            if iterations >- number_of_collection_points:
                break
```